src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.Decimal.cs

```diff
@@ -0,0 +1,259 @@
+
+            _currentDepth |= 1 << 31;
+            _tokenType = JsonTokenType.Number;
+        }
+
+        /// <summary>
+        /// Writes the property name and <see cref="decimal"/> value (as a JSON number) as part of a name/value pair of a JSON object.
+        /// </summary>
+        /// <param name="propertyName">The UTF-8 encoded property name of the JSON object to be written.</param>
+        /// <param name="value">The value to be written as a JSON number as part of the name/value pair.</param>
+        /// <param name="suppressEscaping">If this is set, the writer assumes the property name is properly escaped and skips the escaping step.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown when the specified property name is too large.
+        /// </exception>
+        /// <exception cref="JsonWriterException">
+        /// Thrown if this would result in an invalid JSON to be written (while validation is enabled).
+        /// </exception>
+        public void WriteNumber(ReadOnlySpan<byte> propertyName, decimal value, bool suppressEscaping = false)
+        {
+            JsonWriterHelper.ValidateProperty(ref propertyName);
+
+            if (!suppressEscaping)
+                WriteNumberSuppressFalse(ref propertyName, value);
+            else
+                WriteNumberByOptions(ref propertyName, value);
+
+            _currentDepth |= 1 << 31;
+            _tokenType = JsonTokenType.Number;
+        }
+
+        private void WriteNumberSuppressFalse(ref ReadOnlySpan<char> propertyName, decimal value)
+        {
+            int propertyIdx = JsonWriterHelper.NeedsEscaping(propertyName);
+
+            Debug.Assert(propertyIdx >= -1 && propertyIdx < int.MaxValue / 2);
+
+            if (propertyIdx != -1)
+            {
+                WriteNumberEscapeProperty(ref propertyName, value, propertyIdx);
+            }
+            else
+            {
+                WriteNumberByOptions(ref propertyName, value);
+            }
+        }
+
+        private void WriteNumberSuppressFalse(ref ReadOnlySpan<byte> propertyName, decimal value)
+        {
+            int propertyIdx = JsonWriterHelper.NeedsEscaping(propertyName);
+
+            Debug.Assert(propertyIdx >= -1 && propertyIdx < int.MaxValue / 2);
+
+            if (propertyIdx != -1)
+            {
+                WriteNumberEscapeProperty(ref propertyName, value, propertyIdx);
+            }
+            else
+            {
+                WriteNumberByOptions(ref propertyName, value);
+            }
+        }
+
+        private void WriteNumberEscapeProperty(ref ReadOnlySpan<char> propertyName, decimal value, int firstEscapeIndexProp)
+        {
+            Debug.Assert(int.MaxValue / MaxExpansionFactorWhileEscaping >= propertyName.Length);
+
+            char[] propertyArray = null;
+
+            int length = firstEscapeIndexProp + MaxExpansionFactorWhileEscaping * (propertyName.Length - firstEscapeIndexProp);
```

Comment:
Aside from the fact that we can't pass in `escaped` to `WriteNumberByOptions` unless we use unsafe, that regresses perf by ~20%. What if we use (and re-use) a 256 byte (and 256 char) static array rather than a stackalloc array instead (without the try/finally block)?

Something like:

```C#
private static readonly char[] EscapedPropertyNameChar = new char[256];
private static readonly byte[] EscapedPropertyNameByte = new byte[256];

private unsafe void WriteNumberEscapeProperty(ReadOnlySpan<char> propertyName, long value, int firstEscapeIndexProp)
{
    Debug.Assert(int.MaxValue / 6 >= propertyName.Length);

    char[] rentedBuffer = null;

    ReadOnlySpan<char> escapedPropertyName = GetEscapedProperty(propertyName, ref rentedBuffer, firstEscapeIndexProp);

    WriteNumberByOptions(ref escapedPropertyName, value);

    if (rentedBuffer != null)
    {
        ArrayPool<char>.Shared.Return(rentedBuffer);
    }
}

private static ReadOnlySpan<char> GetEscapedProperty(ReadOnlySpan<char> propertyName, ref char[] rentedBuffer, int firstEscapeIndexProp)
{
    int length = firstEscapeIndexProp + 6 * (propertyName.Length - firstEscapeIndexProp);

    Span<char> span;
    if (length > 256)
    {
        rentedBuffer = ArrayPool<char>.Shared.Rent(length);
        span = rentedBuffer;
    }
    else
    {
        span = EscapedPropertyNameChar;
    }
    JsonWriterHelper.EscapeString(propertyName, span, firstEscapeIndexProp, out int written);
    return span.Slice(0, written);
}
```

Reply:
The arrays would need to be thread-static, or you'd get multithreaded data corruption :smile:.  And at thread-static you're checking initialization every time, you'd want to just rent.

There must be something weird going on in your example, because stackalloc can definitely go into a method without unsafe.

https://github.com/dotnet/corefx/blob/64477348da1ff57a43deb65a4b12d32986ed00bd/src/Common/src/System/Security/Cryptography/Pkcs12Kdf.cs#L129-L165 as one example.

Reply:
> There must be something weird going on in your example, because stackalloc can definitely go into a method without unsafe.

Yes, it can. The question is what happens when you try to pass it to another method. If the method is static, it works fine. If the method is an instance method on a non-ref struct type, ti works fine.

From https://github.com/dotnet/corefx/pull/34425#discussion_r247299400

> Calling static methods by passing a stackalloc span is fine. Calling **instance methods** on a **ref struct** will not work since nothing is stopping us from stashing the stackalloc span as a field. So, the compiler is preventing us from passing a stackalloc span (regardless of whether `propertyName` is passed by ref or not).

```C#
private void WriteStringEscapeValueSafe(int length)
{
    byte[] valueArray = null;

    Span<byte> span = length <= StackallocThreshold ?
        stackalloc byte[length] :
        (valueArray = ArrayPool<byte>.Shared.Rent(length));

    Foo(span); // Fails

    if (valueArray != null)
    {
        ArrayPool<byte>.Shared.Return(valueArray);
    }
}

private void WriteStringEscapeValueUnsafe(int length)
{
    byte[] valueArray = null;
    Span<byte> span;
    if (length > StackallocThreshold)
    {
        valueArray = ArrayPool<byte>.Shared.Rent(length);
        span = valueArray;
    }
    else
    {
        // Cannot create a span directly since the span gets exposed outside this method.
        unsafe
        {
            byte* ptr = stackalloc byte[length];
            span = new Span<byte>(ptr, length);
        }
    }

    Foo(span);

    if (valueArray != null)
    {
        ArrayPool<byte>.Shared.Return(valueArray);
    }
}

private void Foo(Span<byte> value)
{

}
```

Reply:
cc @jaredpar, @agocke - do you guys have any suggestions here to avoid using unsafe code and pass a stackalloc span to an instance method of a ref struct?

Reply:
You should fix #34568 first. 

The depth of the callstacks and number of copies of Span argument that are side-effect of that are your primary problem. It is why passing the Span by `in` or `ref` helped you. There are other overheads associated with the callstack depth. It is quite possible that fixing the callstack depth will get you something that is faster than what you had with the `in` and `ref` hacks.

If you still see cases where you are losing a lot with idiomatic code that passes Span by value, you can ask JIT folks for advice.